keps/sig-apps/3998-job-success-completion-policy/README.md

```diff
@@ -0,0 +1,501 @@
+	// If empty, the default behavior applies - the job is declared as succeeded
+	// only when the number of succeeded pods equals to the completions.
+	// This policy can not be specified more than one.
+	//
+	// This field is alpha-level. To use this field, you must enable the
+	// `JobSuccessPolicy` feature gate (disable by default).
+	// +optional
+	SuccessPolicy *SuccessPolicy
+}
+
+// SuccessPolicy describes when a Job can be declared as succeeded based on the success of some Indexes.
+type SuccessPolicy struct {
+	// Specifies a set of required indexes when .spec.completionMode =
+	// "Indexed" in a text format.
+	// The job is declared as succeeded if a set of indexes are succeeded. 
+	// The list of indexes must come within 0 to `.spec.completions` and
+	// must not contain duplicates. At least one element is required. 
+	// The indexes are represented as decimal integers
+	// separated by commas. The numbers are listed in increasing order. Three or
+	// more consecutive numbers are compressed and represented by the first and
+	// last element of the series, separated by a hyphen.
+	// For example, if the completed indexes are 1, 3, 4, 5 and 7, they are
+	// represented as "1,3-5,7".
+	// 
+	// +optional
+	RequiredIndexes string
+
+	// Specifies the required number of indexes when .spec.completionMode =
+	// "Indexed".
+	// Value can be an absolute number (ex: 5) or a percentage of total indexes
+	// when the job can be declared as succeeded (ex: 50%).
+	// The absolute number is calculated from the percentage by rounding up. 
+	// 
+	// +optional
+	MinSucceededIndexes intstr.IntOrString
+}
+```
+
+Moreover, we validate the following constraints:
+- exactly one of the fields `requiredIndexes` and `minSucceededIndexes` is specified
+  for a requirement.
+- whether the specified indexes in the `requiredIndexes` and
+  the number of indexes in the `minSucceededIndexes` don't exceed the value of `completions`.
+- whether `Indexed` is specified in the `completionMode` field.
+- whether the size of `requiredIndexes` is under 64Ki. 
+
+### Evaluation
+We evaluate the SuccessPolicy after all indexes have been finished.
+And then if the job meets the SuccessPolicy, we add `Completed` condition
+instead of `Failed` condition to `.status.conditions` even if other indexes fail.
+
+If a set of indexes are specified in the `requiredIndexes` field,
+we evaluate `.status.completedIndexes` to see if a set of indexes is there.
+
+If the number of indexes is specified in the `minSucceededIndexes`, 
+we evaluate `.status.succeeded` to see if the value is `minSucceededIndexes` or more.
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+##### Unit tests
+
+- Test cases:
+  - tests for Defaulting and Validating
```

Comment:
What kind of defaulting or validation will you do for this feature?

Reply:
@kannon92 You can check the details in the `Design Details` section.

Reply:
Sorry I missed that!